### PR TITLE
Feature/build package apex classes subfolders

### PR DESCRIPTION
**Pull request by dgalindo001**
## What does this pull request contains? Explain your changes.
Since some releases Salesforce let developers create subfolders under classes metadata. This feature let teams organize the code better. 
This feature branch works to solve this scenario to create package.xml and destructivechanges.xml properly if you have subfolders under classes metadata.

- The pull request has a new class to handle subfolder in classes metadata

---

- [ ] Added for new features.
- [ ] Changed for changes in existing functionality.
- [ ] Deprecated for soon-to-be removed features.
- [ ] Removed for now removed features.
- [ ] Fixed for any bug fixes.
- [ ] Security in case of vulnerabilities.

## Explain your changes
Since some releases Salesforce let developers create subfolders under classes metadata. This feature let teams organize the code better. 
This feature branch works to solve this scenario to create package.xml and destructivechanges.xml properly if you have subfolders under classes metadata.

- The pull request has a new class to handle subfolder in classes metadata

---

## Does this close any currently open issues?

---

- [ ] Jest test to check the fix is applied are added.

## Any particular element to being able to test locally

---

_(Write your answer here.)_

## Any other comments?

---

_(Write your answer here.)_

## Where has this been tested?

---

**Operating System:**MAC

**Yarn version:**v1.22.10

**Node version:**v14.15.1

**sgd version:**Lastest

**git version:**2.24.3

### CODE REVIEW

**Review by mehdisfdc — DISMISSED** [review]

**Comment by dgalindo001** [comment]
@mehdisfdc Do you know how can I merge this PR? I can't see Merge button. Thank you in advance! Regards Dani

**Comment by mehdisfdc** [comment]
> @mehdisfdc Do you know how can I merge this PR? I can't see Merge button. Thank you in advance! Regards Dani

Though this PR look good to me (thank you for contributing!), I will let @scolladon review it as well before merging, since he is the guy behind almost all of the code that in the tool so far.

**Comment by dgalindo001** [comment]
Great! Thank you for your help! @mehdisfdc 
Have a nice day. :)

File: bin/cli
Comment (scolladon) [inline bin/cli:5]: This change should be revert.
Hence it will not be deployable.
To test you need to yarn pack the application.
Test are based on the compiled sources.

File: src/service/subFolderElementHandler.js
Comment (scolladon) [inline src/service/subFolderElementHandler.js:9]: Should be removed, same as the parent

File: src/service/subFolderElementHandler.js
Comment (scolladon) [inline src/service/subFolderElementHandler.js:23]: Maybe in this case we should not call the parent
And deal with the -meta file ourself
```
if (this.metadata[this.type].metaFile === true) {
  this._copyFiles(path.join(this.config.repo, fileName + mc.METAFILE_SUFFIX), path.join(this.config.output, fileName + mc.METAFILE_SUFFIX))
}
```
Something like that

File: src/service/subFolderElementHandler.js
Comment (scolladon) [inline src/service/subFolderElementHandler.js:33]: replace this with : 

```
_getParsedPath() {
    return path.parse(
      this.splittedLine
        .slice(this.splittedLine.length - 1)
        .join(path.sep)
        .replace(mc.META_REGEX, '')
    )
  }
```

Comment (scolladon) [reply src/service/subFolderElementHandler.js:33]: Or : 
```
_getElementName() {
    return 
      this.splittedLine
        .slice(this.splittedLine.length - 1)
        .join(path.sep)
        .replace(mc.META_REGEX, '')
  }
```

Which ever works the best

**Review by scolladon — CHANGES_REQUESTED** [review]
Please look at the changes.
Feel free to ask any question.
Chat is possible.
I'll test it in my environment when the requested changes will be considered

Great contribution ! I like it

**Comment by dgalindo001** [comment]
@scolladon Many thanks for the review! I've made the suggested changes but still have error on build. Could you help me with this?

**Review by scolladon — CHANGES_REQUESTED** [review]
Please remove sfdx-git-delta-v4.0.4.tgz from the elements
You can add sfdx-git-delta-*.tgz to the .gitignore

**Review by scolladon — CHANGES_REQUESTED** [review]
Please remove  sfdx-git-delta-v4.0.4.tgz from the PR :)

**Comment by dgalindo001** [comment]
Sorry @scolladon, my mistake! I think now is fine. 
Thanks! :)

File: src/service/subFolderElementHandler.js
Comment (scolladon) [inline src/service/subFolderElementHandler.js:14]: Should we had the copying of the file itself (and not only the meta file) ? 
```
this._copyFiles(
  path.join(this.config.repo, fileName),
  path.join(this.config.output, fileName)
 )
```

**Review by scolladon — CHANGES_REQUESTED** [review]
Please review my open question.
I will test everything locally to try to have this validated for this we

File: src/service/subFolderElementHandler.js
Comment (dgalindo001) [reply src/service/subFolderElementHandler.js]: No. It's no necessary. 
I'm testing it and it works fine for me only with _getParsedPath method adding the last .replace: 

class SubFolderElementHandler extends StandardHandler {
  _getParsedPath() {
    return path.parse(
      this.splittedLine
        .slice(this.splittedLine.length - 1)
        .join(path.sep)
        .replace(mc.META_REGEX, '')
        **.replace(this.suffixRegex, '')**
    )
  }
}

I've update the PR removing useful override and test review.

**Comment by scolladon** [comment]
The issue in the CI is weird. It is like if the secret to load the coverage to code climate is not accessible.
I tried to update it but it is still not working

**Comment by scolladon** [comment]
@dgalindo001 do you have a link to the documentation explaining this feature on the file structure ?
We will use it in the release note.

**Comment by dgalindo001** [comment]
> @dgalindo001 do you have a link to the documentation explaining this feature on the file structure ?
> We will use it in the release note.

Its a minor change to handle subfolders on apex classes path. It follows the same pattern as others so I added the description to the PR comment:

"Since some releases Salesforce let developers create subfolders under classes metadata. This feature let teams organize the code better.
This feature branch works to solve this scenario to create package.xml and destructivechanges.xml properly if you have subfolders under classes metadata.

The pull request has a new class to handle subfolder in classes metadata"

Where is the file with the documentation? Do you usually add it to README.md?

**Comment by scolladon** [comment]
We try to relate the change at least in the release note in order to understand why we introduce new feature.
I try to find explanation in the Salesforce CLI documentation of this behaviour without success for now.
I thought maybe you could locate it ?

**Comment by dgalindo001** [comment]
I am trying to organize better my apex classes structure so I was looking for information about the possibility of group classes at different folders. 
I found this Salesforce Idea: https://trailblazer.salesforce.com/ideaView?id=08730000000BrmSAAS 
I tested it and I deployed my classes fine. So I checked that this plugin do not support folders at classes. For this reason I created this feature. 

I have not found the feature at Salesforce release note, but the idea lastest comment is over 1 year.

Regards

**Review by scolladon — APPROVED** [review]
Very elegant feature
High quality feature development
thank @dgalindo001 !! Great contribution here

**Pull request merged**
